machine/util/callbacks/early_stopping_callback.py

```diff
@@ -0,0 +1,94 @@
+from machine.util.callbacks import Callback
+
+
+class EarlyStoppingCallback(Callback):
+    """
+    Original callback taken from https://github.com/ncullen93/torchsample
+    Early Stopping to terminate training early under certain conditions
+    """
+
+    def __init__(self,
+                 monitor='eval_losses',
+                 lm_name=None,
+                 min_delta=0,
+                 patience=5,
+                 minimize=True):
+        """
+        EarlyStopping callback to exit the training loop if training or
+        validation loss or metric does not improve by a certain amount
+        for a certain number of epochs
+        Args:
+            monitor : string in {'eval_losses', 'eval_metrics', 'train_losses', 'train_metrics'}
+                whether to monitor train or val loss
+            lm_name: loss or metric name eg. 'Avg NLLoss' or 'Word Accuracy'
+                    If not specified then the first element
+                    in the monitor array is used
+                    (default None)
+            min_delta : float
+                minimum change in monitored value to qualify as improvement.
+                This number should be positive.
+            patience : integer
+                number of epochs to wait for improvment before terminating.
+                the counter be reset after each improvment
+            minimize: minimize quantity, if false then early stopping will maximize
+        """
+        if 'loss' in monitor:
+            self.loss = True
+        elif 'metric' in monitor:
+            self.loss = False
+        else:
+            raise ValueError(
+                "Monitor must be string in {'eval_losses', \
+                'eval_metrics', 'train_losses', 'train_metrics'}")
+
+        self.monitor = monitor
+        self.lm_name = lm_name
+        self.minimize = minimize
+
+        self.min_delta = min_delta
+        self.patience = patience
+        self.wait = 0
+        self.best_lm = 1e-15
```

Comment:
Both `self.wait` and `self.best_lm` are also set in `on_train_begin`, is it necessary to set them also here? Or is this merely to just declare that these attributes exist?

Reply:
Yeah I guess they aren't useful there - I kept them in the case train begin is somehow not called, but since that shouldn't happen - I'll remove them